lib/src/iterable_matchers.dart

```diff
@@ -165,30 +165,39 @@ class _UnorderedMatches extends Matcher {
         return 'has too many elements (${list.length} > ${_expected.length})';
       }
 
-      var matched = new List<bool>.filled(list.length, false);
-      var expectedPosition = 0;
-      for (var expectedMatcher in _expected) {
-        var actualPosition = 0;
-        var gotMatch = false;
-        for (var actualElement in list) {
-          if (!matched[actualPosition]) {
-            if (expectedMatcher.matches(actualElement, {})) {
-              matched[actualPosition] = gotMatch = true;
-              break;
-            }
+      var adjacency = list
+          .map((_) => new List.filled(_expected.length, false, growable: false))
+          .toList(growable: false);
+      for (int v = 0; v < list.length; v++) {
+        for (int m = 0; m < _expected.length; m++) {
+          if (_expected[m].matches(list[v], {})) {
+            adjacency[v][m] = true;
           }
-          ++actualPosition;
         }
-
-        if (!gotMatch) {
+      }
+      // The index into `values` matched with each matcher
+      var matched = new List<int>.filled(_expected.length, -1, growable: false);
```

Comment:
nit: I would slightly prefer `null` to `-1`.

Reply:
Done